pgjdbc/src/test/java/org/postgresql/test/jdbc2/PreparedStatementTest.java

```diff
@@ -1011,4 +1012,39 @@ public void testStatementDescribe() throws SQLException {
     pstmt.close();
   }
 
+  @Test
+  public void testInappropriateStatementSharing() throws SQLException {
+    PreparedStatement ps = con.prepareStatement("SELECT ?::timestamp");
+    try {
+      Timestamp ts = new Timestamp(1474997614836L);
+      // Since PreparedStatement isn't cached immediately, we need to some warm up
+      for (int i = 0; i < 3; ++i) {
+        ResultSet rs;
+
+        // Flip statement to use Oid.DATE
+        ps.setNull(1, Types.DATE);
+        rs = ps.executeQuery();
+        try {
+          assertTrue(rs.next());
+          assertNull(rs.getObject(1));
+        } finally {
+          rs.close();
+        }
+
+        // Flop statement to use Oid.UNSPECIFIED
+        ps.setTimestamp(1, ts);
+        rs = ps.executeQuery();
+        try {
+          assertTrue(rs.next());
+          // If an inappropriate caching occurred, then we'll get the data narrowing (TIMESTAMP -> DATE)
+          assertEquals(ts, rs.getObject(1));
```

Comment:
The comment above this line should probably be moved into message argument of assertEquals. That would make error message much cleaner.